In [52]:
import numpy as np
import pandas as pd
import pickle


#from train import satisfaction

In [53]:
random_forest = pickle.load(open('../data/variables_pkl/resultado_random_forest.pkl', 'rb'))
svc = pickle.load(open('../data/variables_pkl/resultado_svc.pkl', 'rb'))
log_reg = pickle.load(open('../data/variables_pkl/resultado_log_reg.pkl', 'rb'))
log_reg_rf = pickle.load(open('../data/variables_pkl/resultado_log_reg_rf.pkl', 'rb'))
knn = pickle.load(open('../data/variables_pkl/resultado_knn.pkl', 'rb'))
xgb = pickle.load(open('../data/variables_pkl/resultado_xgb.pkl', 'rb'))
gbc = pickle.load(open('../data/variables_pkl/resultado_gbc.pkl', 'rb'))


In [54]:
resultado = pd.DataFrame({log_reg, log_reg_rf, random_forest, knn, svc , gbc, xgb},columns=['Modelo', 'Score'])
resultado.sort_values(by='Score', ignore_index=True, ascending=False)

Modelo     Score
0                  XGBoosting  0.986582
1               Random Forest  0.974142
2  GradientBoostingClassifier  0.961218
3               resultado_KNN  0.931490
4     log_regcon selección RF  0.911170
5                         SVC  0.910432
6                     log_reg  0.908290

 Dados los resultados de todos los modelos el mejor es el XGBoosting con un score de 0.986582, pero para poder conseguir el proposito de detectar el grado de satisfacción del cliente funciona mejor el random forest.

 Lo implemento en el archivo train.py, recoge datos nuevos, ya que no han sido usados para entrenar los modelos del df_test de la carpeta de data/processed.

In [55]:
random_forest_model = pickle.load(open('../models/random_forest.pkl', 'rb'))
data_prueba = pd.read_csv('../data/processed/df_test.csv').sample(frac=1)
respuesta = data_prueba['satisfaction'].iloc[:1].values
print(f'respuesta encuesta: {respuesta}')
pred_proba = random_forest_model.predict_proba(data_prueba.iloc[:1:, :18:].values)
print(pred_proba)
grado_sast = list(pred_proba[0])
print(grado_sast[0])


if respuesta == 0:
    print('grado de no satisfacción es del: {:.0f}%'.format(grado_sast[0]*100))
else:
    print('grado de satisfacción es del: {:.0f}%'.format(grado_sast[0]*100))


respuesta encuesta: [1]
[[0.03535845 0.96464155]]
0.03535844990889505
grado de satisfacción es del: 4%


In [41]:
grado_sast

[0.9581620622116505, 0.0418379377883496]

In [35]:
def satisfaction():
    try:
        random_forest_model = pickle.load(open('../models/random_forest.pkl', 'rb'))
        data_prueba = pd.read_csv('../data/processed/df_test.csv').sample(frac=1)
        respuesta = data_prueba.satisfaction.iloc[:1].min()
        print(f'respuesta encuesta: {respuesta}')
        pred_proba = random_forest_model.predict_proba(data_prueba.iloc[:1:, :18:].values)
        #print(pred_proba)
        grado_sast = list(pred_proba[0])
        #print(grado_sast[0])


        if respuesta == 0:
            print('grado de no satisfacción es del: {:.0f}%'.format(grado_sast[0]*100))
        else:
            print('grado de satisfacción es del: {:.0f}%'.format(grado_sast[0]*100))
    except UserWarning:
        print('')


In [36]:
satisfaction()

respuesta encuesta: 1
grado de satisfacción es del: 16%


/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/sklearn/base.py:451: UserWarning: X does not have valid feature names, but RandomForestClassifier was fitted with feature names
  "X does not have valid feature names, but"
